library/src/cache/patch_manager.rs

```diff
@@ -0,0 +1,883 @@
+        );
+        assert_eq!(format!("{:?}", patch_manager), "ManagePatches");
+    }
+
+    #[test]
+    fn patch_manager_is_debug() {
+        let temp_dir = TempDir::new("patch_manager").unwrap();
+        let patch_manager = PatchManager::with_root_dir(temp_dir.path().to_owned());
+        let expected_str = format!(
+            "PatchManager {{ root_dir: \"{}\", patches_state: PatchesState {{ last_booted_patch: None, next_boot_patch: None, highest_seen_patch_number: None }} }}",
+            temp_dir.path().display()
+        );
+        assert_eq!(format!("{:?}", patch_manager), expected_str);
+    }
+}
+
+#[cfg(test)]
+mod add_patch_tests {
+    use super::*;
+    use std::path::Path;
+    use tempdir::TempDir;
+
+    #[test]
+    fn errs_if_file_path_does_not_exist() {
+        let mut manager = PatchManager::manager_for_test(&TempDir::new("patch_manager").unwrap());
+        assert!(manager
+            .add_patch(1, Path::new("/path/to/file/that/does/not/exist"))
+            .is_err());
+    }
+
+    #[test]
+    fn adds_patch_successfully() {
+        let patch_number = 1;
+        let patch_file_contents = "patch contents";
+        let temp_dir = TempDir::new("patch_manager").unwrap();
+        let mut manager = PatchManager::with_root_dir(temp_dir.path().to_owned());
+
+        let file_path = &temp_dir.path().join("patch1.vmcode");
+        std::fs::write(file_path, patch_file_contents).unwrap();
+
+        assert!(manager
+            .add_patch(patch_number, Path::new(file_path))
+            .is_ok());
+
+        assert_eq!(
+            manager.patches_state.next_boot_patch,
+            Some(PatchMetadata {
+                number: patch_number,
+                size: patch_file_contents.len() as u64
+            })
+        );
+        assert!(!file_path.exists());
+        assert_eq!(manager.highest_seen_patch_number(), Some(patch_number));
+    }
+
+    #[test]
+    fn does_not_set_higher_highest_seen_patch_number_if_added_patch_is_lower() -> Result<()> {
+        let patch_file_contents = "patch contents";
+
+        let temp_dir = TempDir::new("patch_manager")?;
+        let mut manager = PatchManager::manager_for_test(&temp_dir);
+        assert!(manager.highest_seen_patch_number().is_none());
+
+        // Add patch 1
+        let file_path = &temp_dir.path().join("patch.vmcode");
+        std::fs::write(file_path, patch_file_contents)?;
+        assert!(manager.add_patch(1, file_path).is_ok());
+        assert_eq!(manager.highest_seen_patch_number(), Some(1));
+
+        // Add patch 4, expect 4 to be the highest patch number we've seen
+        let file_path = &temp_dir.path().join("patch.vmcode");
+        std::fs::write(file_path, patch_file_contents)?;
+        assert!(manager.add_patch(4, file_path).is_ok());
+        assert_eq!(manager.highest_seen_patch_number(), Some(4));
+
+        // Add patch 3, expect 4 to still be the highest patch number we've seen
```

Comment:
Does this imply that the way we'll end up implementing rollback is by copying a patch to a new patch number?

Reply:
I'm not sure what specifically you mean by rollback. Are you referring to the case where a new patch fails to boot, or a developer-initiated rollback feature that we'll build in the future?

Reply:
I was imagining how a future feature might work and realizing that clients are going to store state about what patch they've seen and will refuse to go backwards to a patch.  So if we implement rollback we'll need to "copy" the patch and assign it a new number.

Reply:
>  if we implement rollback we'll need to "copy" the patch and assign it a new number.

Yeah, this is how I imagined it would work – take the artifacts from our rollback target patch and make it patch N+1